clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1130,8 +1130,96 @@ struct BitTest {
 
   static BitTest decodeBitTestBuiltin(unsigned BuiltinID);
 };
+
+// Returns the first convergence entry/loop/anchor instruction found in |BB|.
+// std::nullopt otherwise.
+std::optional<llvm::IntrinsicInst *> getConvergenceToken(llvm::BasicBlock *BB) {
+  for (auto &I : *BB) {
```

Comment:
It feels wrong that you would need to do a scan of a block to find this

Reply:
Yep, but I'm not sure I have any other option outside from some kind of caching(either attribute of a BB, or BB->I map). If it's this kind of alternative you are thinking about, maybe this should be left in a subsequent PR as an optimization?

Reply:
This kind of map will be needed in more places as we continue to introduces uses of these convergence tokens. I have a simple Analysis in my upstreaming pipeline that produces such a map, but it's really very simple. You can wait for that, or just feel free to start your own.

Reply:
Good to know! I'd be in favor to implement this map once we have a merged use case.
-> This way, we can have an independent NFC PR which focuses on this optimization/refactoring by adding either an analysis or the bits in `llvm::BasicBlock` to quickly fetch those.